### DISCUSSION TITLE
Conversion functions for custom data type cannot be found.

### DISCUSSION THREAD

**Discussion by ThePythonator** [nlohmann/json#3383] [Q&A] [answered]
Hi, I'm using [this](https://github.com/sgorsten/linalg) linear algebra header (specifically the `vec` struct), and I need to save and load these from JSON files.
I've written a simple conversion function for the 2D vector, but whenever I call `get<linalg::vec<T,2>()` or `get_to(v)`, I receive this error:
```
'nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>>::get_to': no matching overloaded function found
```

Here are the conversion functions. They're in the global namespace, so I'm certain that they should be able to be located by the json header, and I've double checked that the signatures are as recommended in your README.
```cpp
template<class T>
void to_json(json& j, const linalg::vec<T, 2>& v) {
	j = json{
		{v.x, v.y}
	};
}

template<class T>
void from_json(const json& j, linalg::vec<T, 2>& v) {
	j.at(0).get_to(v.x);
	j.at(1).get_to(v.y);
}
```

Have I missing something/made a mistake with these functions, or might the definition of `vec` prevent it from being serialized in this manner? I've tried functions similar to these with my own custom data types and they've worked fine.

Thanks for your time.

**Comment by nlohmann** [comment, marked as answer]
You need to define them in the namespace of the type.

**Reply by gregmarr** [reply]
If you can't put them in the type's namespace because you don't control it, see here: https://github.com/nlohmann/json#how-do-i-convert-third-party-types

**Reply by ThePythonator** [reply]
Thank you for your help, putting those functions inside the `linalg` namespace fixed the compiler error, but I now get 'unresolved external symbol' linker errors whenever I  (implicitly) use the `to_json` method.

For example, in the conversion for another class:
```cpp
void to_json(json& j, const Component& c) {
	j = json{
		{"type", c.get_type()},
		{"nodes", c.get_nodes()},
		//{"position", c.get_position()} // Uncommenting this line causes a linker error
	};
}
```

`c.get_position()` returns a `vec<double,2>` type.

**Reply by gregmarr** [reply]
What is the full linker error?

**Reply by ThePythonator** [reply]
`LNK2019	unresolved external symbol "void __cdecl linalg::to_json<double>(class nlohmann::basic_json<class std::map,class std::vector,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,bool,__int64,unsigned __int64,double,class std::allocator,struct nlohmann::adl_serializer,class std::vector<unsigned char,class std::allocator<unsigned char> > > &,struct linalg::vec<double,2> const &)" (??$to_json@N@linalg@@YAXAEAV?$basic_json@Vmap@std@@Vvector@2@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@2@_N_J_KNVallocator@2@Uadl_serializer@nlohmann@@V?$vector@EV?$allocator@E@std@@@2@@nlohmann@@AEBU?$vec@N$01@0@@Z) referenced in function "void __cdecl to_json(class nlohmann::basic_json<class std::map,class std::vector,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,bool,__int64,unsigned __int64,double,class std::allocator,struct nlohmann::adl_serializer,class std::vector<unsigned char,class std::allocator<unsigned char> > > &,class Component const &)" (?to_json@@YAXAEAV?$basic_json@Vmap@std@@Vvector@2@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@2@_N_J_KNVallocator@2@Uadl_serializer@nlohmann@@V?$vector@EV?$allocator@E@std@@@2@@nlohmann@@AEBVComponent@@@Z)`

**Reply by ThePythonator** [reply]
Is it possible that the template is messing with it because the definition is in the `.cpp` file, and the compiler may not be aware of `nlohmann::json` calling the `double` version of the function?

**Reply by gregmarr** [reply]
You have this in a header file that is included before your `to_json` function?

    namespace linalg {
        template<class T>
        void to_json(json& j, const vec<T, 2>& v) {
	        j = json{
		        {v.x, v.y}
	        };
        }

        template<class T>
        void from_json(const json& j, vec<T, 2>& v) {
	        j.at(0).get_to(v.x);
	        j.at(1).get_to(v.y);
        }
    };

**Reply by ThePythonator** [reply]
Yes. The basic structure is:
`linalg_extras.hpp`, `linalg_extras.cpp` - with the `to_json` and `from_json` functions for `linalg::vec<T, 2>`.
`Component.hpp`, `Component.cpp` - definition of `Component` class along with `to_json` and `from_json` functions for it.

**Reply by ThePythonator** [reply]
I've tried without the template (using `linalg::vec<double, 2>` instead), and it works.

I'm not sure why templating doesn't work even with specialization, but I'll use the non-template method for now since I only wanted to implement templates in case I needed more types of `vec` in the future.

Thank you both for your help!

**Reply by gregmarr** [reply]
The template function bodies need to be in the header file that is included by `Component.cpp`.  Putting them in a `.cpp` file that is not included will not work.

**Reply by ThePythonator** [reply]
Thanks, I tried putting the bodies into the header and it works with the templates now, which is great. Thanks for your time!

**Reply by gregmarr** [reply]
You're welcome.
